[SCISPARK 159] Refactor MCSUtils, separate function responsibilities.

rahulpalamuttam opened the pull request:
Addresses issue #159

Refactored MCSUtils to separate some of the responsibilities of the functions.

Initially the chain of calls went get_node_data -> extract_masked_data -> netcdfWrite.
However, these function calls were called within each other.

The PR introduces a new function writeEdgeComponentsToNetCDF()
The individual functions are decoupled from each other and called within writeEdgeComponentsToNetCDF(). 

get_node_data has been renamed to getMCSNodes, which looks up the edge nodes in the MCSNodeMap and returns them as a tuple.

extract_masked_data returns a tuple of nodeID and the grid object as an m2 Array.

netcdfWrite has been renamed to writeNodeToNetCDF.

rahulpalamuttam committed:
[SCISPARK 159] Refactor MCSUtils, separate function responsibilities.

rahulpalamuttam commented on src/main/scala/org/dia/algorithms/mcc/MCSUtils.scala:
```diff
@@ -31,125 +31,155 @@ import org.dia.tensors.AbstractTensor
+   * @param edge       the edge object which is composed of the nodes to look up in the MCSNodeMap
+   * @param MCSNodeMap mutable.HashMap[String, MCCNode] representing the map of each node metadata
+   */
+  def getMCSNodes(edge: MCCEdge, MCSNodeMap: mutable.HashMap[String, MCCNode]): (MCCNode, MCCNode) = {
```
@sujen there's some questions I had here.
1. Instead of returning a tuple of nodes - could we just return a populated edge object?
2. The node object in the mccedge seems to just be a shell and the MCSNodemap contains the nodes with all the metadata.  (Just verifying)

rahulpalamuttam commented on src/main/scala/org/dia/algorithms/mcc/MCSUtils.scala:
```diff
@@ -31,125 +31,155 @@ import org.dia.tensors.AbstractTensor
+  def extract_masked_data(thisNode: MCCNode,
+                          lats: Array[Double],
+                          lons: Array[Double],
+                          tightestBox: Boolean): (String, ArrayInt.D2) = {
```
I wonder if there's a better way to return the ID and Array, instead of using a tuple.

kwhitehall commented on src/main/scala/org/dia/algorithms/mcc/MCSUtils.scala:
```diff
@@ -31,125 +31,155 @@ import org.dia.tensors.AbstractTensor
+   * @param lons        Array[Double] of the longitudes to be used
+   * @param tightestBox Boolean to use tightest box around data.
+   * @param MCSNodeMap  mutable.HashMap[String, MCCNode] representing the map of each node metadata
+   */
```
fix @param ordering to that in the function definition.

kwhitehall commented:
@rahulpalamuttam a couple items on this PR
1. the name writeEdgeComponentsToNetCDF() doesn't fit with the nomenclature, and introduces yet another term IMO. It would be better to call this something like writeEdgeNodesToNetCDF() or writeEdgeCloudElementsToNetCDF()
2. In general, when contributing to the code, please avoid copy and pasting large chunks of code under new functions and deleting the old function copied from. It gives the illusion that large changes have been made, when really a renaming of the function definition may have occurred or a few lines in the original function added. An e.g. in this PR is writeNodeToNetCDF() and netcdf_write(), where only renaming has occurred. 
[EDIT] for more mature code bases, it would be prudent to enter a line indicating that the function is being replaced. @chrismattmann  @lewismc can you please direct here on the best opensource practice?

rahulpalamuttam commented:
@kwhitehall @lewismc @chrismattmann 
The reason why it looks like it has been copy pasted, is because I also changed the indentation.
If you take a look MCSUtils in master there are two levels of indentation in the class code.

Please take a look at what's in master right now:
https://github.com/SciSpark/SciSpark/blob/master/src/main/scala/org/dia/algorithms/mcc/MCSUtils.scala

and compare it to what's in the PR

https://github.com/rahulpalamuttam/SciSpark/blob/39af4c59dc632b1b98c35e36d39f5198389f37d2/src/main/scala/org/dia/algorithms/mcc/MCSUtils.scala

Master clearly has an excessive level of indentation.
Sorry that I didn't specify the indent change in the PR.

kwhitehall commented:
Ack'ed.

rahulpalamuttam committed:
[SCISPARK 159] Change function parameter ordering and name

rahulpalamuttam commented:
@kwhitehall @chrismattmann @lewismc
The new commit addresses your comments.
1. The parameters are re-ordered for the functions
2. The function writeEdgeComponentsToNetCDF has been renamed to writeEdgeNodesToNetCDF

Are you okay with the state of the PR right now? The apparent copy pasting is a result of indent change on the file to comply with the 2-space indent format we use throughout the project.
We didn't catch this in the initial PR #150  that added the file, because the scalastyle-config.xml didn't have the indent checker enabled. I filed a new issue #167 for it so the style checker can catch indent errors in future PR's.

kwhitehall commented:
Sweet @rahulpalamuttam. LGTM. Merging. Thanks for addressing the spaces and opening an issue for the checker to handle it by default.
